panic in schema store

### What did you do?

```
goroutine 1022 [running]:
go.uber.org/zap/zapcore.CheckWriteAction.OnWrite(0x3?, 0x3?, {0x0?, 0x0?, 0xc003402960?})
        go.uber.org/zap@v1.27.0/zapcore/entry.go:196 +0x54
go.uber.org/zap/zapcore.(*CheckedEntry).Write(0xc003522f70, {0xc003539800, 0x3, 0x3})
        go.uber.org/zap@v1.27.0/zapcore/entry.go:262 +0x24e
go.uber.org/zap.(*Logger).Panic(0x0?, {0x5978c7a?, 0xc003539800?}, {0xc003539800, 0x3, 0x3})
        go.uber.org/zap@v1.27.0/logger.go:285 +0x51
github.com/pingcap/log.Panic({0x5978c7a?, 0xaa?}, {0xc003539800?, 0xc003271780?, 0x65667c77dbc0008?})
        github.com/pingcap/log@v1.1.1-0.20241212030209-7e3ff8601a2a/global.go:54 +0x85
github.com/pingcap/ticdc/pkg/eventservice.(*eventBroker).addDispatcher(0xc0016e3a40, {0x641b0c0, 0xc00351c8e8})
        github.com/pingcap/ticdc/pkg/eventservice/event_broker.go:868 +0x11f9
github.com/pingcap/ticdc/pkg/eventservice.(*eventService).registerDispatcher(0xc0041b03c0, {0x63e0320, 0xc0041b05a0}, {0x641b0c0, 0xc00351c8e8})
        github.com/pingcap/ticdc/pkg/eventservice/event_service.go:145 +0xaf
github.com/pingcap/ticdc/pkg/eventservice.(*eventService).Run(0xc0041b03c0, {0x63e0320, 0xc0041b05a0})
        github.com/pingcap/ticdc/pkg/eventservice/event_service.go:97 +0x191
github.com/pingcap/ticdc/server.(*server).Run.(*server).Run.func2.func3()
        github.com/pingcap/ticdc/server/server.go:237 +0x2fa
golang.org/x/sync/errgroup.(*Group).Go.func1()
        golang.org/x/sync@v0.10.0/errgroup/errgroup.go:78 +0x50
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 1
        golang.org/x/sync@v0.10.0/errgroup/errgroup.go:75 +0x96
panic: get table info from schemaStore failed
```
From https://asktug.com/t/topic/1040166/16

### What did you expect to see?

_No response_

### What did you see instead?

Nan

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
(paste TiCDC version here)
```

--- a/logservice/schemastore/disk_format.go
+++ b/logservice/schemastore/disk_format.go
@@ -84,6 +84,18 @@ func tableInfoKey(ts uint64, tableID int64) ([]byte, error) {
 	return buf.Bytes(), nil
 }
 
+func partitionInfoKey(ts uint64, partitionID int64) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	buf.WriteString(snapshotPartitionKeyPrefix)
+	if err := binary.Write(buf, binary.BigEndian, ts); err != nil {
+		return nil, err
+	}
+	if err := binary.Write(buf, binary.BigEndian, partitionID); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
 func ddlJobKey(ts uint64) ([]byte, error) {
 	buf := new(bytes.Buffer)
 	buf.WriteString(ddlKeyPrefix)
@@ -486,12 +498,18 @@ func addSchemaInfoToBatch(batch *pebble.Batch, ts uint64, info *model.DBInfo) {
 	batch.Set(schemaKey, schemaValue, pebble.NoSync)
 }
 
-func addTableInfoToBatch(batch *pebble.Batch, ts uint64, dbInfo *model.DBInfo, tableInfoValue []byte) (int64, string) {
-	tbNameInfo := model.TableNameInfo{}
-	if err := json.Unmarshal(tableInfoValue, &tbNameInfo); err != nil {
+func addTableInfoToBatch(
+	batch *pebble.Batch,
+	ts uint64,
+	dbInfo *model.DBInfo,
+	tableInfoValue []byte,
+) (int64, string, []int64) {
+	tableInfo := model.TableInfo{}
+	if err := json.Unmarshal(tableInfoValue, &tableInfo); err != nil {
 		log.Fatal("unmarshal table info failed", zap.Error(err))
 	}
-	tableKey, err := tableInfoKey(ts, tbNameInfo.ID)
+	// write table info to batch
+	tableKey, err := tableInfoKey(ts, tableInfo.ID)
 	if err != nil {
 		log.Fatal("generate table key failed", zap.Error(err))
 	}
@@ -505,16 +523,33 @@ func addTableInfoToBatch(batch *pebble.Batch, ts uint64, dbInfo *model.DBInfo, t
 		log.Fatal("marshal table info entry failed", zap.Error(err))
 	}
 	batch.Set(tableKey, tableInfoEntryValue, pebble.NoSync)
-	return tbNameInfo.ID, tbNameInfo.Name.O
+
+	// write partition info to batch if the table is a partition table
+	var partitionIDs []int64
+	if tableInfo.Partition != nil {
+		for _, partition := range tableInfo.Partition.Definitions {
+			partitionKey, err := partitionInfoKey(ts, partition.ID)
+			if err != nil {
+				log.Fatal("generate partition key failed", zap.Error(err))
+			}
+			valueBuf := new(bytes.Buffer)
+			if err := binary.Write(valueBuf, binary.BigEndian, tableInfo.ID); err != nil {
+				log.Fatal("generate partition value failed", zap.Error(err))
+			}
+			batch.Set(partitionKey, valueBuf.Bytes(), pebble.NoSync)
+			partitionIDs = append(partitionIDs, partition.ID)
+		}
+	}
+	return tableInfo.ID, tableInfo.Name.O, partitionIDs
 }
 
-// writeSchemaSnapshotAndMeta write database info and table info to disks.
-func writeSchemaSnapshotAndMeta(
+// persistSchemaSnapshot write database/table/partition info to disks.
+func persistSchemaSnapshot(
 	db *pebble.DB,
 	tiStore kv.Storage,
 	snapTs uint64,
-	needTableInfo bool,
-) (map[int64]*BasicDatabaseInfo, map[int64]*BasicTableInfo, error) {
+	collectMetaInfo bool,
+) (map[int64]*BasicDatabaseInfo, map[int64]*BasicTableInfo, map[int64]BasicPartitionInfo, error) {
 	meta := getSnapshotMeta(tiStore, snapTs)
 	start := time.Now()
 	dbInfos, err := meta.ListDatabases()
@@ -523,56 +558,62 @@ func writeSchemaSnapshotAndMeta(
 	}
 
 	var databaseMap map[int64]*BasicDatabaseInfo
-	var tablesInKVSnap map[int64]*BasicTableInfo
-	if needTableInfo {
+	var tableMap map[int64]*BasicTableInfo
+	var partitionMap map[int64]BasicPartitionInfo
+	if collectMetaInfo {
 		databaseMap = make(map[int64]*BasicDatabaseInfo)
-		tablesInKVSnap = make(map[int64]*BasicTableInfo)
+		tableMap = make(map[int64]*BasicTableInfo)
+		partitionMap = make(map[int64]BasicPartitionInfo)
 	}
 	for _, dbInfo := range dbInfos {
 		if filter.IsSysSchema(dbInfo.Name.O) {
 			continue
 		}
 		batch := db.NewBatch()
-
 		addSchemaInfoToBatch(batch, snapTs, dbInfo)
-
 		rawTables, err := meta.GetMetasByDBID(dbInfo.ID)
 		if err != nil {
 			log.Fatal("get tables failed", zap.Error(err))
 		}
-		var tables map[int64]bool
-		if needTableInfo {
-			tables = make(map[int64]bool)
+		var tablesInDB map[int64]bool
+		if collectMetaInfo {
+			tablesInDB = make(map[int64]bool)
 		}
 		for _, rawTable := range rawTables {
 			if !isTableRawKey(rawTable.Field) {
 				continue
 			}
-			tableID, tableName := addTableInfoToBatch(batch, snapTs, dbInfo, rawTable.Value)
-			if needTableInfo {
-				tablesInKVSnap[tableID] = &BasicTableInfo{
+			tableID, tableName, partitionIDs := addTableInfoToBatch(batch, snapTs, dbInfo, rawTable.Value)
+			if collectMetaInfo {
+				tableMap[tableID] = &BasicTableInfo{
 					SchemaID: dbInfo.ID,
 					Name:     tableName,
 				}
-				tables[tableID] = true
+				tablesInDB[tableID] = true
+				if len(partitionIDs) > 0 {
+					partitionMap[tableID] = make(BasicPartitionInfo)
+					for _, partitionID := range partitionIDs {
+						partitionMap[tableID].AddPartitionID(partitionID)
+					}
+				}
 			}
 			// 8M is arbitrary, we can adjust it later
 			if batch.Len() >= 8*1024*1024 {
 				if err := batch.Commit(pebble.NoSync); err != nil {
-					return nil, nil, err
+					return nil, nil, nil, err
 				}
 				batch = db.NewBatch()
 			}
 		}
-		if needTableInfo {
+		if collectMetaInfo {
 			databaseInfo := &BasicDatabaseInfo{
 				Name:   dbInfo.Name.O,
-				Tables: tables,
+				Tables: tablesInDB,
 			}
 			databaseMap[dbInfo.ID] = databaseInfo
 		}
 		if err := batch.Commit(pebble.NoSync); err != nil {
-			return nil, nil, err
+			return nil, nil, nil, err
 		}
 	}
 
@@ -580,7 +621,7 @@ func writeSchemaSnapshotAndMeta(
 
 	log.Info("finish write schema snapshot",
 		zap.Any("duration", time.Since(start).Seconds()))
-	return databaseMap, tablesInKVSnap, nil
+	return databaseMap, tableMap, partitionMap, nil
 }
 
 func cleanObsoleteData(db *pebble.DB, oldGcTs uint64, gcTs uint64) {

--- a/logservice/schemastore/disk_format.go
+++ b/logservice/schemastore/disk_format.go
@@ -372,31 +372,70 @@ func loadAndApplyDDLHistory(
 	return tablesDDLHistory, tableTriggerDDLHistory, nil
 }
 
-func readTableInfoInKVSnap(snap *pebble.Snapshot, tableID int64, version uint64) *common.TableInfo {
-	targetKey, err := tableInfoKey(version, tableID)
+// if tableID is a physical partition id, return the logic table id of it
+func tryReadLogicalTableID(snap *pebble.Snapshot, tableID int64, version uint64) int64 {
+	key, err := partitionInfoKey(version, tableID)
 	if err != nil {
-		log.Fatal("generate table info failed", zap.Error(err))
+		return 0
 	}
-	value, closer, err := snap.Get(targetKey)
+	val, closer, err := snap.Get(key)
 	if err == pebble.ErrNotFound {
-		return nil
+		return 0
 	}
 	if err != nil {
-		log.Fatal("get table info failed", zap.Error(err))
+		log.Fatal("read partition meta failed", zap.Error(err))
+		return 0
 	}
 	defer closer.Close()
 
-	var table_info_entry PersistedTableInfoEntry
-	if _, err := table_info_entry.UnmarshalMsg(value); err != nil {
-		log.Fatal("unmarshal table info entry failed", zap.Error(err))
+	if len(val) != 8 {
+		log.Fatal("invalid meta value length", zap.Int("len", len(val)))
+		return 0
 	}
+	return int64(binary.BigEndian.Uint64(val))
+}
 
-	tableInfo := &model.TableInfo{}
-	err = json.Unmarshal(table_info_entry.TableInfoValue, tableInfo)
-	if err != nil {
-		log.Fatal("unmarshal table info failed", zap.Error(err))
+func readTableInfoInKVSnap(snap *pebble.Snapshot, tableID int64, version uint64) *common.TableInfo {
+	readRawTableInfo := func(targetTableID int64) (string, *model.TableInfo) {
+		targetKey, err := tableInfoKey(version, targetTableID)
+		if err != nil {
+			log.Fatal("generate table info failed", zap.Error(err))
+		}
+		value, closer, err := snap.Get(targetKey)
+		if err == pebble.ErrNotFound {
+			return "", nil
+		}
+		if err != nil {
+			log.Fatal("get table info failed", zap.Error(err))
+		}
+		defer closer.Close()
+
+		var table_info_entry PersistedTableInfoEntry
+		if _, err := table_info_entry.UnmarshalMsg(value); err != nil {
+			log.Fatal("unmarshal table info entry failed", zap.Error(err))
+		}
+
+		tableInfo := &model.TableInfo{}
+		err = json.Unmarshal(table_info_entry.TableInfoValue, tableInfo)
+		if err != nil {
+			log.Fatal("unmarshal table info failed", zap.Error(err))
+		}
+		return table_info_entry.SchemaName, tableInfo
 	}
-	return common.WrapTableInfo(table_info_entry.SchemaName, tableInfo)
+	schemaName, tableInfo := readRawTableInfo(tableID)
+	if tableInfo == nil {
+		// check whether it a physical partition id
+		logicalTableID := tryReadLogicalTableID(snap, tableID, version)
+		if logicalTableID != 0 {
+			schemaName, tableInfo = readRawTableInfo(logicalTableID)
+			if tableInfo == nil {
+				return nil
+			}
+		} else {
+			return nil
+		}
+	}
+	return common.WrapTableInfo(schemaName, tableInfo)
 }
 
 func unmarshalPersistedDDLEvent(value []byte) PersistedDDLEvent {

--- a/logservice/schemastore/disk_format.go
+++ b/logservice/schemastore/disk_format.go
@@ -46,11 +46,11 @@ import (
 //     and we will pull ddl job from `resolved_ts` at restart if the current gc ts is smaller than resolved_ts.
 
 const (
-	snapshotSchemaKeyPrefix = "ss_"
-	snapshotTableKeyPrefix  = "st_"
+	snapshotSchemaKeyPrefix    = "ss_"
+	snapshotTableKeyPrefix     = "st_"
+	snapshotPartitionKeyPrefix = "sp_"
+	ddlKeyPrefix               = "ds_"
 )
-
-const ddlKeyPrefix = "ds_"
 
 func gcTsKey() []byte {
 	return []byte("gc")

--- a/logservice/schemastore/disk_format.go
+++ b/logservice/schemastore/disk_format.go
@@ -474,7 +474,7 @@ func isTableRawKey(key []byte) bool {
 	return strings.HasPrefix(string(key), mTablePrefix)
 }
 
-func writeSchemaInfoToBatch(batch *pebble.Batch, ts uint64, info *model.DBInfo) {
+func addSchemaInfoToBatch(batch *pebble.Batch, ts uint64, info *model.DBInfo) {
 	schemaKey, err := schemaInfoKey(ts, info.ID)
 	if err != nil {
 		log.Fatal("generate schema key failed", zap.Error(err))
@@ -486,7 +486,7 @@ func writeSchemaInfoToBatch(batch *pebble.Batch, ts uint64, info *model.DBInfo) 
 	batch.Set(schemaKey, schemaValue, pebble.NoSync)
 }
 
-func writeTableInfoToBatch(batch *pebble.Batch, ts uint64, dbInfo *model.DBInfo, tableInfoValue []byte) (int64, string) {
+func addTableInfoToBatch(batch *pebble.Batch, ts uint64, dbInfo *model.DBInfo, tableInfoValue []byte) (int64, string) {
 	tbNameInfo := model.TableNameInfo{}
 	if err := json.Unmarshal(tableInfoValue, &tbNameInfo); err != nil {
 		log.Fatal("unmarshal table info failed", zap.Error(err))
@@ -508,6 +508,7 @@ func writeTableInfoToBatch(batch *pebble.Batch, ts uint64, dbInfo *model.DBInfo,
 	return tbNameInfo.ID, tbNameInfo.Name.O
 }
 
+// writeSchemaSnapshotAndMeta write database info and table info to disks.
 func writeSchemaSnapshotAndMeta(
 	db *pebble.DB,
 	tiStore kv.Storage,
@@ -533,7 +534,7 @@ func writeSchemaSnapshotAndMeta(
 		}
 		batch := db.NewBatch()
 
-		writeSchemaInfoToBatch(batch, snapTs, dbInfo)
+		addSchemaInfoToBatch(batch, snapTs, dbInfo)
 
 		rawTables, err := meta.GetMetasByDBID(dbInfo.ID)
 		if err != nil {
@@ -547,7 +548,7 @@ func writeSchemaSnapshotAndMeta(
 			if !isTableRawKey(rawTable.Field) {
 				continue
 			}
-			tableID, tableName := writeTableInfoToBatch(batch, snapTs, dbInfo, rawTable.Value)
+			tableID, tableName := addTableInfoToBatch(batch, snapTs, dbInfo, rawTable.Value)
 			if needTableInfo {
 				tablesInKVSnap[tableID] = &BasicTableInfo{
 					SchemaID: dbInfo.ID,

--- a/logservice/schemastore/disk_format.go
+++ b/logservice/schemastore/disk_format.go
@@ -632,7 +632,7 @@ func persistSchemaSnapshot(
 				if len(partitionIDs) > 0 {
 					partitionMap[tableID] = make(BasicPartitionInfo)
 					for _, partitionID := range partitionIDs {
-						partitionMap[tableID].AddPartitionID(partitionID)
+						partitionMap[tableID].AddPartitionIDs(partitionID)
 					}
 				}
 			}

--- a/logservice/schemastore/multi_version.go
+++ b/logservice/schemastore/multi_version.go
@@ -98,7 +98,8 @@ func (e *TableDeletedError) Error() string {
 	return "table deleted"
 }
 
-// return the table info with the largest version <= ts
+// if ts < deleteVersion, return the table info with the largest version <= ts
+// if ts >= deleteVersion, return nil table info and TableDeletedError
 func (v *versionedTableInfoStore) getTableInfo(ts uint64) (*common.TableInfo, error) {
 	v.mu.Lock()
 	defer v.mu.Unlock()

--- a/logservice/schemastore/multi_version_test_utils.go
+++ b/logservice/schemastore/multi_version_test_utils.go
@@ -74,24 +74,15 @@ func buildCreatePartitionTableEventForTest(schemaID, tableID int64, schemaName, 
 		SchemaName: schemaName,
 		TableName:  tableName,
 		TableInfo: &model.TableInfo{
-			ID:   tableID,
-			Name: pmodel.NewCIStr(tableName),
-			Partition: &model.PartitionInfo{
-				Definitions: partitionDefinitions,
-				Enable:      true,
-			},
+			ID:        tableID,
+			Name:      pmodel.NewCIStr(tableName),
+			Partition: buildPartitionDefinitionsForTest(partitionIDs),
 		},
 		FinishedTs: finishedTs,
 	}
 }
 
 func buildDropPartitionTableEventForTest(schemaID, tableID int64, schemaName, tableName string, partitionIDs []int64, finishedTs uint64) *PersistedDDLEvent {
-	partitionDefinitions := make([]model.PartitionDefinition, 0, len(partitionIDs))
-	for _, partitionID := range partitionIDs {
-		partitionDefinitions = append(partitionDefinitions, model.PartitionDefinition{
-			ID: partitionID,
-		})
-	}
 	return &PersistedDDLEvent{
 		Type:       byte(model.ActionDropTable),
 		SchemaID:   schemaID,
@@ -99,12 +90,9 @@ func buildDropPartitionTableEventForTest(schemaID, tableID int64, schemaName, ta
 		SchemaName: schemaName,
 		TableName:  tableName,
 		TableInfo: &model.TableInfo{
-			ID:   tableID,
-			Name: pmodel.NewCIStr(tableName),
-			Partition: &model.PartitionInfo{
-				Definitions: partitionDefinitions,
-				Enable:      true,
-			},
+			ID:        tableID,
+			Name:      pmodel.NewCIStr(tableName),
+			Partition: buildPartitionDefinitionsForTest(partitionIDs),
 		},
 		FinishedTs: finishedTs,
 	}
@@ -131,12 +119,6 @@ func buildTruncatePartitionTableEventForTest(
 	schemaName, tableName string,
 	newPartitionIDs []int64, finishedTs uint64,
 ) *PersistedDDLEvent {
-	partitionDefinitions := make([]model.PartitionDefinition, 0, len(newPartitionIDs))
-	for _, partitionID := range newPartitionIDs {
-		partitionDefinitions = append(partitionDefinitions, model.PartitionDefinition{
-			ID: partitionID,
-		})
-	}
 	return &PersistedDDLEvent{
 		Type:       byte(model.ActionTruncateTable),
 		SchemaID:   schemaID,
@@ -144,12 +126,9 @@ func buildTruncatePartitionTableEventForTest(
 		SchemaName: schemaName,
 		TableName:  tableName,
 		TableInfo: &model.TableInfo{
-			ID:   newTableID,
-			Name: pmodel.NewCIStr(tableName),
-			Partition: &model.PartitionInfo{
-				Definitions: partitionDefinitions,
-				Enable:      true,
-			},
+			ID:        newTableID,
+			Name:      pmodel.NewCIStr(tableName),
+			Partition: buildPartitionDefinitionsForTest(newPartitionIDs),
 		},
 		FinishedTs: finishedTs,
 	}
@@ -178,12 +157,6 @@ func buildExchangePartitionTableEventForTest(
 	normalSchemaName, normalTableName, partitionSchemaName, partitionTableName string,
 	oldPartitionIDs, newPartitionIDs []int64, finishedTs uint64,
 ) *PersistedDDLEvent {
-	partitionDefinitions := make([]model.PartitionDefinition, 0, len(newPartitionIDs))
-	for _, partitionID := range newPartitionIDs {
-		partitionDefinitions = append(partitionDefinitions, model.PartitionDefinition{
-			ID: partitionID,
-		})
-	}
 	return &PersistedDDLEvent{
 		Type:            byte(model.ActionExchangeTablePartition),
 		SchemaID:        normalSchemaID,
@@ -195,12 +168,9 @@ func buildExchangePartitionTableEventForTest(
 		ExtraSchemaName: partitionSchemaName,
 		ExtraTableName:  partitionTableName,
 		TableInfo: &model.TableInfo{
-			ID:   partitionTableID,
-			Name: pmodel.NewCIStr(partitionTableName),
-			Partition: &model.PartitionInfo{
-				Definitions: partitionDefinitions,
-				Enable:      true,
-			},
+			ID:        partitionTableID,
+			Name:      pmodel.NewCIStr(partitionTableName),
+			Partition: buildPartitionDefinitionsForTest(newPartitionIDs),
 		},
 		ExtraTableInfo: common.WrapTableInfo(normalSchemaName, &model.TableInfo{
 			ID:   normalTableID,

--- a/logservice/schemastore/multi_version_test_utils.go
+++ b/logservice/schemastore/multi_version_test_utils.go
@@ -211,7 +211,7 @@ func buildExchangePartitionTableEventForTest(
 	}
 }
 
-func buildPartitionDefinitions(partitionIDs []int64) *model.PartitionInfo {
+func buildPartitionDefinitionsForTest(partitionIDs []int64) *model.PartitionInfo {
 	partitionDefinitions := make([]model.PartitionDefinition, 0, len(partitionIDs))
 	for _, partitionID := range partitionIDs {
 		partitionDefinitions = append(partitionDefinitions, model.PartitionDefinition{

--- a/logservice/schemastore/multi_version_test_utils.go
+++ b/logservice/schemastore/multi_version_test_utils.go
@@ -210,3 +210,16 @@ func buildExchangePartitionTableEventForTest(
 		FinishedTs:     finishedTs,
 	}
 }
+
+func buildPartitionDefinitions(partitionIDs []int64) *model.PartitionInfo {
+	partitionDefinitions := make([]model.PartitionDefinition, 0, len(partitionIDs))
+	for _, partitionID := range partitionIDs {
+		partitionDefinitions = append(partitionDefinitions, model.PartitionDefinition{
+			ID: partitionID,
+		})
+	}
+	return &model.PartitionInfo{
+		Definitions: partitionDefinitions,
+		Enable:      true,
+	}
+}

--- a/logservice/schemastore/persist_storage.go
+++ b/logservice/schemastore/persist_storage.go
@@ -197,7 +197,7 @@ func (p *persistentStorage) initializeFromKVStorage(dbPath string, storage kv.St
 		zap.Uint64("snapTs", gcTs))
 
 	var err error
-	if p.databaseMap, p.tableMap, err = writeSchemaSnapshotAndMeta(p.db, storage, gcTs, true); err != nil {
+	if p.databaseMap, p.tableMap, p.partitionMap, err = persistSchemaSnapshot(p.db, storage, gcTs, true); err != nil {
 		// TODO: retry
 		log.Fatal("fail to initialize from kv snapshot")
 	}
@@ -210,6 +210,7 @@ func (p *persistentStorage) initializeFromKVStorage(dbPath string, storage kv.St
 	log.Info("schema store initialize from kv storage done",
 		zap.Int("databaseMapLen", len(p.databaseMap)),
 		zap.Int("tableMapLen", len(p.tableMap)),
+		zap.Int("partitionMapLen", len(p.partitionMap)),
 		zap.Any("duration(s)", time.Since(now).Seconds()))
 }
 
@@ -540,7 +541,7 @@ func (p *persistentStorage) doGc(gcTs uint64) error {
 	}
 
 	start := time.Now()
-	_, _, err := writeSchemaSnapshotAndMeta(p.db, p.kvStorage, gcTs, false)
+	_, _, _, err := persistSchemaSnapshot(p.db, p.kvStorage, gcTs, false)
 	if err != nil {
 		log.Warn("fail to write kv snapshot during gc",
 			zap.Uint64("gcTs", gcTs))

--- a/logservice/schemastore/persist_storage_ddl_handlers.go
+++ b/logservice/schemastore/persist_storage_ddl_handlers.go
@@ -1016,9 +1016,7 @@ func updateSchemaMetadataForNewTableDDL(args updateSchemaMetadataFuncArgs) {
 	}
 	if isPartitionTable(args.event.TableInfo) {
 		partitionInfo := make(BasicPartitionInfo)
-		for _, id := range getAllPartitionIDs(args.event.TableInfo) {
-			partitionInfo[id] = nil
-		}
+		partitionInfo.AddPartitionIDs(getAllPartitionIDs(args.event.TableInfo)...)
 		args.partitionMap[tableID] = partitionInfo
 	}
 }
@@ -1049,9 +1047,7 @@ func updateSchemaMetadataForTruncateTable(args updateSchemaMetadataFuncArgs) {
 	if isPartitionTable(args.event.TableInfo) {
 		delete(args.partitionMap, oldTableID)
 		partitionInfo := make(BasicPartitionInfo)
-		for _, id := range getAllPartitionIDs(args.event.TableInfo) {
-			partitionInfo[id] = nil
-		}
+		partitionInfo.AddPartitionIDs(getAllPartitionIDs(args.event.TableInfo)...)
 		args.partitionMap[newTableID] = partitionInfo
 	}
 }
@@ -1142,9 +1138,7 @@ func updateSchemaMetadataForCreateTables(args updateSchemaMetadataFuncArgs) {
 		}
 		if isPartitionTable(info) {
 			partitionInfo := make(BasicPartitionInfo)
-			for _, id := range getAllPartitionIDs(info) {
-				partitionInfo[id] = nil
-			}
+			partitionInfo.AddPartitionIDs((getAllPartitionIDs(info))...)
 			args.partitionMap[info.ID] = partitionInfo
 		}
 	}
@@ -1154,13 +1148,9 @@ func updateSchemaMetadataForReorganizePartition(args updateSchemaMetadataFuncArg
 	tableID := args.event.TableID
 	physicalIDs := getAllPartitionIDs(args.event.TableInfo)
 	droppedIDs := getDroppedIDs(args.event.PrevPartitions, physicalIDs)
-	for _, id := range droppedIDs {
-		delete(args.partitionMap[tableID], id)
-	}
+	args.partitionMap[tableID].RemovePartitionIDs(droppedIDs...)
 	newCreatedIDs := getCreatedIDs(args.event.PrevPartitions, physicalIDs)
-	for _, id := range newCreatedIDs {
-		args.partitionMap[tableID][id] = nil
-	}
+	args.partitionMap[tableID].AddPartitionIDs(newCreatedIDs...)
 }
 
 func updateSchemaMetadataForAlterTablePartitioning(args updateSchemaMetadataFuncArgs) {
@@ -1178,9 +1168,7 @@ func updateSchemaMetadataForAlterTablePartitioning(args updateSchemaMetadataFunc
 		Name:     args.event.TableName,
 	}
 	args.partitionMap[newTableID] = make(BasicPartitionInfo)
-	for _, id := range getAllPartitionIDs(args.event.TableInfo) {
-		args.partitionMap[newTableID][id] = nil
-	}
+	args.partitionMap[newTableID].AddPartitionIDs(getAllPartitionIDs(args.event.TableInfo)...)
 }
 
 func updateSchemaMetadataForRemovePartitioning(args updateSchemaMetadataFuncArgs) {

--- a/logservice/schemastore/persist_storage_test.go
+++ b/logservice/schemastore/persist_storage_test.go
@@ -368,11 +368,23 @@ func TestApplyDDLJobs(t *testing.T) {
 		// test partition table related ddl
 		{
 			"partition table",
-			nil,
+			[]mockDBInfo{
+				{
+					dbInfo: &model.DBInfo{
+						ID:   100,
+						Name: pmodel.NewCIStr("test"),
+					},
+					tables: []*model.TableInfo{
+						{
+							ID:        200,
+							Name:      pmodel.NewCIStr("t1"),
+							Partition: buildPartitionDefinitionsForTest([]int64{201, 202, 203}),
+						},
+					},
+				},
+			},
 			func() []*model.Job {
 				return []*model.Job{
-					buildCreateSchemaJobForTest(100, "test", 1000),                                           // create schema 100
-					buildCreatePartitionTableJobForTest(100, 200, "t1", []int64{201, 202, 203}, 1010),        // create partition table 200
 					buildTruncatePartitionTableJobForTest(100, 200, 300, "t1", []int64{204, 205, 206}, 1020), // truncate partition table 200 to 300
 					buildAddPartitionJobForTest(100, 300, "t1", []int64{204, 205, 206, 207}, 1030),           // add partition 207
 					buildDropPartitionJobForTest(100, 300, "t1", []int64{205, 206, 207}, 1040),               // drop partition 204
@@ -401,16 +413,16 @@ func TestApplyDDLJobs(t *testing.T) {
 				},
 			},
 			map[int64][]uint64{
-				201: {1010, 1020},
-				202: {1010, 1020},
-				203: {1010, 1020},
+				201: {1020},
+				202: {1020},
+				203: {1020},
 				204: {1020, 1030, 1040},
 				205: {1020, 1030, 1040, 1050},
 				206: {1020, 1030, 1040, 1050},
 				207: {1030, 1040, 1050},
 				208: {1050},
 			},
-			[]uint64{1000, 1010, 1020, 1030, 1040, 1050},
+			[]uint64{1020, 1030, 1040, 1050},
 			[]PhysicalTableQueryTestCase{
 				{
 					snapTs: 1010,
@@ -516,11 +528,29 @@ func TestApplyDDLJobs(t *testing.T) {
 					limit:   1,
 					result: []commonEvent.DDLEvent{
 						{
-							Type:       byte(model.ActionCreateSchema),
-							FinishedTs: 1000,
+							Type:       byte(model.ActionTruncateTable),
+							FinishedTs: 1020,
 							BlockedTables: &commonEvent.InfluencedTables{
 								InfluenceType: commonEvent.InfluenceTypeNormal,
-								TableIDs:      []int64{0},
+								TableIDs:      []int64{0, 201, 202, 203},
+							},
+							NeedDroppedTables: &commonEvent.InfluencedTables{
+								InfluenceType: commonEvent.InfluenceTypeNormal,
+								TableIDs:      []int64{201, 202, 203},
+							},
+							NeedAddedTables: []commonEvent.Table{
+								{
+									SchemaID: 100,
+									TableID:  204,
+								},
+								{
+									SchemaID: 100,
+									TableID:  205,
+								},
+								{
+									SchemaID: 100,
+									TableID:  206,
+								},
 							},
 						},
 					},
@@ -529,36 +559,6 @@ func TestApplyDDLJobs(t *testing.T) {
 					startTs: 1000,
 					limit:   10,
 					result: []commonEvent.DDLEvent{
-						{
-							Type:       byte(model.ActionCreateTable),
-							FinishedTs: 1010,
-							BlockedTables: &commonEvent.InfluencedTables{
-								InfluenceType: commonEvent.InfluenceTypeNormal,
-								TableIDs:      []int64{0},
-							},
-							NeedAddedTables: []commonEvent.Table{
-								{
-									SchemaID: 100,
-									TableID:  201,
-								},
-								{
-									SchemaID: 100,
-									TableID:  202,
-								},
-								{
-									SchemaID: 100,
-									TableID:  203,
-								},
-							},
-							TableNameChange: &commonEvent.TableNameChange{
-								AddName: []commonEvent.SchemaTableName{
-									{
-										SchemaName: "test",
-										TableName:  "t1",
-									},
-								},
-							},
-						},
 						{
 							Type:       byte(model.ActionTruncateTable),
 							FinishedTs: 1020,
@@ -2244,7 +2244,7 @@ func TestRegisterTable(t *testing.T) {
 						{
 							ID:        102,
 							Name:      pmodel.NewCIStr("t1"),
-							Partition: buildPartitionDefinitions([]int64{201, 202, 203}),
+							Partition: buildPartitionDefinitionsForTest([]int64{201, 202, 203}),
 						},
 					},
 				},

--- a/logservice/schemastore/persist_storage_test.go
+++ b/logservice/schemastore/persist_storage_test.go
@@ -2259,17 +2259,17 @@ func TestRegisterTable(t *testing.T) {
 				{
 					tableID: 201,
 					snapTs:  1029,
-					name:    "t4",
+					name:    "t1",
 				},
 				{
 					tableID: 202,
 					snapTs:  1029,
-					name:    "t4",
+					name:    "t1",
 				},
 				{
 					tableID: 203,
 					snapTs:  1029,
-					name:    "t4",
+					name:    "t1",
 				},
 				{
 					tableID: 201,

--- a/logservice/schemastore/persist_storage_test.go
+++ b/logservice/schemastore/persist_storage_test.go
@@ -2121,12 +2121,16 @@ func TestRegisterTable(t *testing.T) {
 		name    string
 	}
 	testCases := []struct {
-		initailDBInfos []mockDBInfo // tables in initailDBInfos will be registered before apply ddl
+		name           string
+		initialDBInfos []mockDBInfo
 		ddlJobs        []*model.Job
+		preDDLTables   []int64 // tables registered before apply ddl
+		postDDLTables  []int64 // tables registered after apply ddl
 		queryCases     []QueryTableInfoTestCase
 	}{
 		{
-			initailDBInfos: []mockDBInfo{
+			name: "rename table",
+			initialDBInfos: []mockDBInfo{
 				{
 					dbInfo: &model.DBInfo{
 						ID:   50,
@@ -2142,12 +2146,10 @@ func TestRegisterTable(t *testing.T) {
 			},
 			ddlJobs: func() []*model.Job {
 				return []*model.Job{
-					buildRenameTableJobForTest(50, 99, "t2", 1000, nil),                              // rename table 99 to t2
-					buildCreateTableJobForTest(50, 100, "t3", 1010),                                  // create table 100
-					buildTruncateTableJobForTest(50, 100, 101, "t3", 1020),                           // truncate table 100 to 101
-					buildCreatePartitionTableJobForTest(50, 102, "t4", []int64{201, 202, 203}, 1030), // create partition table 102
+					buildRenameTableJobForTest(50, 99, "t2", 1000, nil), // rename table 99 to t2
 				}
 			}(),
+			postDDLTables: []int64{99},
 			queryCases: []QueryTableInfoTestCase{
 				{
 					tableID: 99,
@@ -2159,6 +2161,26 @@ func TestRegisterTable(t *testing.T) {
 					snapTs:  1000,
 					name:    "t2",
 				},
+			},
+		},
+		{
+			name: "truncate table",
+			initialDBInfos: []mockDBInfo{
+				{
+					dbInfo: &model.DBInfo{
+						ID:   50,
+						Name: pmodel.NewCIStr("test"),
+					},
+				},
+			},
+			ddlJobs: func() []*model.Job {
+				return []*model.Job{
+					buildCreateTableJobForTest(50, 100, "t3", 1010),        // create table 100
+					buildTruncateTableJobForTest(50, 100, 101, "t3", 1020), // truncate table 100 to 101
+				}
+			}(),
+			postDDLTables: []int64{100, 101},
+			queryCases: []QueryTableInfoTestCase{
 				{
 					tableID: 100,
 					snapTs:  1010,
@@ -2169,6 +2191,25 @@ func TestRegisterTable(t *testing.T) {
 					snapTs:  1020,
 					name:    "t3",
 				},
+			},
+		},
+		{
+			name: "create partition table",
+			initialDBInfos: []mockDBInfo{
+				{
+					dbInfo: &model.DBInfo{
+						ID:   50,
+						Name: pmodel.NewCIStr("test"),
+					},
+				},
+			},
+			ddlJobs: func() []*model.Job {
+				return []*model.Job{
+					buildCreatePartitionTableJobForTest(50, 102, "t4", []int64{201, 202, 203}, 1030), // create partition table 102
+				}
+			}(),
+			postDDLTables: []int64{201, 202, 203},
+			queryCases: []QueryTableInfoTestCase{
 				{
 					tableID: 201,
 					snapTs:  1030,
@@ -2188,24 +2229,28 @@ func TestRegisterTable(t *testing.T) {
 		},
 	}
 	for _, tt := range testCases {
-		dbPath := fmt.Sprintf("/tmp/testdb-%s", t.Name())
-		pStorage := newPersistentStorageForTest(dbPath, tt.initailDBInfos)
-		for _, db := range tt.initailDBInfos {
-			for _, table := range db.tables {
-				pStorage.registerTable(table.ID, 0) // second arguments is not important
+		t.Run(tt.name, func(t *testing.T) {
+			dbPath := fmt.Sprintf("/tmp/testdb-%s", t.Name())
+			pStorage := newPersistentStorageForTest(dbPath, tt.initialDBInfos)
+			for _, tableID := range tt.preDDLTables {
+				err := pStorage.registerTable(tableID, 0) // second arguments is not important
+				require.Nil(t, err)
 			}
-		}
-		for _, job := range tt.ddlJobs {
-			err := pStorage.handleDDLJob(job)
-			require.Nil(t, err)
-		}
-		for _, testCase := range tt.queryCases {
-			pStorage.registerTable(testCase.tableID, 0) // second arguments is not important
-			tableInfo, err := pStorage.getTableInfo(testCase.tableID, testCase.snapTs)
-			require.Nil(t, err)
-			require.Equal(t, testCase.name, tableInfo.TableName.Table)
-		}
-		pStorage.close()
+			for _, job := range tt.ddlJobs {
+				err := pStorage.handleDDLJob(job)
+				require.Nil(t, err)
+			}
+			for _, tableID := range tt.postDDLTables {
+				err := pStorage.registerTable(tableID, 0) // second arguments is not important
+				require.Nil(t, err)
+			}
+			for _, testCase := range tt.queryCases {
+				tableInfo, err := pStorage.getTableInfo(testCase.tableID, testCase.snapTs)
+				require.Nil(t, err)
+				require.Equal(t, testCase.name, tableInfo.TableName.Table)
+			}
+			pStorage.close()
+		})
 	}
 }
 

--- a/logservice/schemastore/persist_storage_test.go
+++ b/logservice/schemastore/persist_storage_test.go
@@ -2119,14 +2119,19 @@ func TestRegisterTable(t *testing.T) {
 		tableID int64
 		snapTs  uint64
 		name    string
+		deleted bool
 	}
 	testCases := []struct {
 		name           string
 		initialDBInfos []mockDBInfo
 		ddlJobs        []*model.Job
-		preDDLTables   []int64 // tables registered before apply ddl
-		postDDLTables  []int64 // tables registered after apply ddl
-		queryCases     []QueryTableInfoTestCase
+		// tables registered before apply ddl
+		// used for test to apply online ddl jobs
+		preDDLTables []int64
+		// tables registered after apply ddl
+		// used for test to load and apply ddls from disk
+		postDDLTables []int64
+		queryCases    []QueryTableInfoTestCase
 	}{
 		{
 			name: "rename table",
@@ -2227,6 +2232,52 @@ func TestRegisterTable(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "drop partition table",
+			initialDBInfos: []mockDBInfo{
+				{
+					dbInfo: &model.DBInfo{
+						ID:   50,
+						Name: pmodel.NewCIStr("test"),
+					},
+					tables: []*model.TableInfo{
+						{
+							ID:        102,
+							Name:      pmodel.NewCIStr("t1"),
+							Partition: buildPartitionDefinitions([]int64{201, 202, 203}),
+						},
+					},
+				},
+			},
+			preDDLTables: []int64{201, 202, 203},
+			ddlJobs: func() []*model.Job {
+				return []*model.Job{
+					buildDropPartitionTableJobForTest(50, 102, "t4", []int64{201, 202, 203}, 1030), // drop partition table 102
+				}
+			}(),
+			queryCases: []QueryTableInfoTestCase{
+				{
+					tableID: 201,
+					snapTs:  1029,
+					name:    "t4",
+				},
+				{
+					tableID: 202,
+					snapTs:  1029,
+					name:    "t4",
+				},
+				{
+					tableID: 203,
+					snapTs:  1029,
+					name:    "t4",
+				},
+				{
+					tableID: 201,
+					snapTs:  1030,
+					deleted: true,
+				},
+			},
+		},
 	}
 	for _, tt := range testCases {
 		t.Run(tt.name, func(t *testing.T) {
@@ -2246,8 +2297,15 @@ func TestRegisterTable(t *testing.T) {
 			}
 			for _, testCase := range tt.queryCases {
 				tableInfo, err := pStorage.getTableInfo(testCase.tableID, testCase.snapTs)
-				require.Nil(t, err)
-				require.Equal(t, testCase.name, tableInfo.TableName.Table)
+				if testCase.deleted {
+					require.Nil(t, tableInfo)
+					if _, ok := err.(*TableDeletedError); !ok {
+						t.Error("expect TableDeletedError, but got", err)
+					}
+				} else {
+					require.Nil(t, err)
+					require.Equal(t, testCase.name, tableInfo.TableName.Table)
+				}
 			}
 			pStorage.close()
 		})

--- a/logservice/schemastore/persist_storage_test_utils.go
+++ b/logservice/schemastore/persist_storage_test_utils.go
@@ -126,13 +126,13 @@ func mockWriteKVSnapOnDisk(db *pebble.DB, snapTs uint64, dbInfos []mockDBInfo) {
 	batch := db.NewBatch()
 	defer batch.Close()
 	for _, dbInfo := range dbInfos {
-		writeSchemaInfoToBatch(batch, snapTs, dbInfo.dbInfo)
+		addSchemaInfoToBatch(batch, snapTs, dbInfo.dbInfo)
 		for _, tableInfo := range dbInfo.tables {
 			tableInfoValue, err := json.Marshal(tableInfo)
 			if err != nil {
 				log.Panic("marshal table info fail", zap.Error(err))
 			}
-			writeTableInfoToBatch(batch, snapTs, dbInfo.dbInfo, tableInfoValue)
+			addTableInfoToBatch(batch, snapTs, dbInfo.dbInfo, tableInfoValue)
 		}
 	}
 	if err := batch.Commit(pebble.NoSync); err != nil {

--- a/logservice/schemastore/types.go
+++ b/logservice/schemastore/types.go
@@ -116,6 +116,14 @@ type BasicTableInfo struct {
 
 type BasicPartitionInfo map[int64]interface{}
 
+func (b BasicPartitionInfo) AddPartitionID(id int64) {
+	b[id] = nil
+}
+
+func (b BasicPartitionInfo) RemovePartitionID(id int64) {
+	delete(b, id)
+}
+
 //msgp:ignore DDLJobWithCommitTs
 type DDLJobWithCommitTs struct {
 	Job *model.Job

--- a/logservice/schemastore/types.go
+++ b/logservice/schemastore/types.go
@@ -116,12 +116,16 @@ type BasicTableInfo struct {
 
 type BasicPartitionInfo map[int64]interface{}
 
-func (b BasicPartitionInfo) AddPartitionID(id int64) {
-	b[id] = nil
+func (info BasicPartitionInfo) AddPartitionIDs(ids ...int64) {
+	for _, id := range ids {
+		info[id] = nil
+	}
 }
 
-func (b BasicPartitionInfo) RemovePartitionID(id int64) {
-	delete(b, id)
+func (info BasicPartitionInfo) RemovePartitionIDs(ids ...int64) {
+	for _, id := range ids {
+		delete(info, id)
+	}
 }
 
 //msgp:ignore DDLJobWithCommitTs

